### PR TITLE
Support TOML configurations for DM

### LINKED ISSUE #4311: Support TOML configurations for DM
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently, all the components in TidbCluster CR have supported [TOML](https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/types.go#L429) configurations, however, for DM, we still use the structured [config](https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/types.go#L2132), which is not flexible enough, for example, if DM supports a new configuration, we also have to update code and release it before users can set that config for DM, e.g. https://github.com/pingcap/tidb-operator/pull/4299.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by better0332**
### What problem does this PR solve?

  Support TOML configurations for DM.

  Closes #4311

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/dm_config_wraper.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	}
+}
+
+type MasterConfigWraper struct {
+	*config.GenericConfig `json:",inline"`
+}
+
+type WorkerConfigWraper struct {
+	*config.GenericConfig `json:",inline"`
+}
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/dm_config_wraper.go:47]: Is it OK to use one config wrapper?

File: pkg/manager/member/dm_master_member_manager.go [hunk 2]
```diff
@@ -730,19 +730,19 @@ func getNewMasterSetForDMCluster(dc *v1alpha1.DMCluster, cm *corev1.ConfigMap) (
...
-		config.SSLCA = pointer.StringPtr(path.Join(dmMasterClusterCertPath, tlsSecretRootCAKey))
-		config.SSLCert = pointer.StringPtr(path.Join(dmMasterClusterCertPath, corev1.TLSCertKey))
-		config.SSLKey = pointer.StringPtr(path.Join(dmMasterClusterCertPath, corev1.TLSPrivateKeyKey))
+		config.Set("ssl-ca", path.Join(dmMasterClusterCertPath, tlsSecretRootCAKey))
+		config.Set("ssl-cert", path.Join(dmMasterClusterCertPath, corev1.TLSCertKey))
+		config.Set("ssl-key", path.Join(dmMasterClusterCertPath, corev1.TLSPrivateKeyKey))
 	}
 
-	confText, err := MarshalTOML(config)
+	confText, err := config.MarshalTOML()
```
Comment (DanielZhangQD) [inline pkg/manager/member/dm_master_member_manager.go:745]: We need to make sure no rolling update is triggered for the existing DM cluster with configuration set explicitly after upgrading the TiDB Operator.
Please refer to https://github.com/pingcap/tidb-operator/blob/master/pkg/manager/member/tikv_member_manager.go#L248.

**Comment by DanielZhangQD** [comment]
Please verify the following cases and update the result in the PR:
1. Deploy TiDB Operator v1.2.6 + DM with configurations set explicitly in the CR, TLS is disabled, after upgrading the TiDB Operator, no rolling update for DM, DM works normally, update the configuration, DM is rolling updated.
2. Deploy TiDB Operator v1.2.6 + DM with configurations set explicitly in the CR, TLS is enabled, after upgrading the TiDB Operator, no rolling update for DM, DM works normally, update the configuration, DM is rolling updated.
3. Deploy new TiDB Operator, DM with configurations set explicitly in the CR, DM works normally, update the configuration, DM is rolling updated.

File: pkg/manager/member/dm_master_member_manager.go (hunk 2 above)
Comment (better0332) [reply pkg/manager/member/dm_master_member_manager.go:745]: dm has used mngerutils.UpdateConfigMapIfNeed() too

**Comment by better0332** [comment]
> Please verify the following cases and update the result in the PR:
> 
> 1. Deploy TiDB Operator v1.2.6 + DM with configurations set explicitly in the CR, TLS is disabled, after upgrading the TiDB Operator, no rolling update for DM, DM works normally, update the configuration, DM is rolling updated.
> 2. Deploy TiDB Operator v1.2.6 + DM with configurations set explicitly in the CR, TLS is enabled, after upgrading the TiDB Operator, no rolling update for DM, DM works normally, update the configuration, DM is rolling updated.
> 3. Deploy new TiDB Operator, DM with configurations set explicitly in the CR, DM works normally, update the configuration, DM is rolling updated.

I have test above, everything is ok.

File: pkg/apis/pingcap/v1alpha1/defaulting/dmcluster_test.go [hunk 3]
```diff
@@ -27,22 +27,9 @@ func TestSetDMSpecDefault(t *testing.T) {
 	SetDMClusterDefault(dc)
 	g.Expect(dc.Spec.Master.Config).Should(BeNil())
 
-	dc = newDMCluster()
-	rpcTimeoutStr := "40s"
-	dc.Spec.Master.Config = &v1alpha1.MasterConfig{
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/defaulting/dmcluster_test.go:32]: Why remove this? Can we update it according to the new config format?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by KanShiori — APPROVED** [review]

**Pull request merged**